### PR TITLE
Update packaging and workflows

### PR DESCRIPTION
**Pull request by wsp-mbuckley**
Describe Changes
---
Updated packaging settings to newer standard using `pyproject.toml`.
Added standard CAF github workflows from CAF.toolkit.

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

**Comment by wsp-mbuckley** [comment]
@BenTaylor-TfN I think this covers everything from a packaging point of view and it seems to install fine from git or from a local editable. I have noticed it fails installing for Python 3.12 because it can't build a pandas wheel, maybe it can't find one on pypi for 3.12 yet?

I'm going to test the packaging with the xy_plots branch next

File: pyproject.toml [hunk 1]
```diff
@@ -1,28 +1,84 @@
...
+dependencies = [
+  "numpy>=1.19.0, <2.0.0",
+  "pandas>=1.4.0, <2.0.0",
+  "matplotlib>=3.8.2, <4.0.0",
+]
+
+[project.optional-dependencies]
+lint = [
+  "black>=22.6.0",
+  "mypy>=1.0.0",
```
Comment (BenTaylor-TfN) [inline pyproject.toml:28]:
```suggestion
  "black>=22.6.0",
  "isort>=5.12.0",
  "mypy>=1.0.0",
```

File: requirements_dev.txt [hunk 2]
```diff
@@ -1,20 +1,19 @@
 # # Packaging controls
-setuptools>=61.2.0
-versioneer>=0.22
-packaging>=21.3
+setuptools>=64.0.0
```
Comment (BenTaylor-TfN) [inline requirements_dev.txt:2]: I think you can get away with not including this, as if you build the package locally this is installed using the build requirements in pyproject.toml
```suggestion
```

**Review by BenTaylor-TfN — CHANGES_REQUESTED** [review]
Mostly there, just a couple minor changes suggested.

Couple of general questions/comments too:
- Does this package not support Python 3.9? What makes it non-supporting?
- It's probably worth also removing the tox files, if they're not being used anymore by anyone too.

**Comment by wsp-mbuckley** [comment]
> Mostly there, just a couple minor changes suggested.
> 
> Couple of general questions/comments too:
> 
> * Does this package not support Python 3.9? What makes it non-supporting?
> * It's probably worth also removing the tox files, if they're not being used anymore by anyone too.

Done all the changes you requested, including removing the tox file. Are for Python 3.9 the xy_plots branch uses the `Concatenate` type annotation from the typing module which was added in 3.10.

File: requirements_dev.txt [hunk 3]
```diff
@@ -1,20 +1,18 @@
...
 
 # Testing
-pytest~=7.1.2
-pytest-cov~=2.12.1
-pytest-xdist~=2.5.0
-tox~=3.24.3
+pytest>=7.4.0
+pytest-cov>=4.1.0
+pytest-xdist>=3.3.1
+tox>=3.24.3
```
Comment (BenTaylor-TfN) [inline requirements_dev.txt:15]:
```suggestion
```

File: pyproject.toml [hunk 4]
```diff
@@ -1,28 +1,84 @@
...
+classifiers = [
+  "Programming Language :: Python :: 3 :: Only",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+]
+dependencies = [
+  "numpy>=1.19.0, <2.0.0",
+  "pandas>=1.4.0, <2.0.0",
+  "matplotlib>=3.8.2, <4.0.0",
```
[the remark this answers is not in the document]
Comment (BenTaylor-TfN) [reply pyproject.toml:22]: It's acting as if you've made this change, but the file looks the same. Could you double check?

**Review by BenTaylor-TfN — APPROVED** [review]
